feat(new sink): add AWS Simple Notification Service `aws_sns` sink

wochinge opened the pull request:
**Proposed Changes**
- add support for [AWS Simple Notification Service (SNS sink)](https://aws.amazon.com/sns/)
- This is useful to distribute logs in a fan out fashion (as opposed to AWS SQS)
  - SNS is very similar to SQS so should be quite simple to maintain this alongside SQS.
- separated out common code from SQS and SNS

Closes: #18142

wochinge commented on src/sinks/aws_sns/integration_tests.rs:
```diff
@@ -0,0 +1,114 @@
+
+    sleep(Duration::from_secs(1)).await;
+
+    // Somewhat lost how to receive the published message
```
seems the simplest option to test the roundtrip would be 
1. Creating a sqs queue which subscribes to the topic
2. Receive the message from the SQS queue

wochinge commented on src/sinks/aws_sns/service.rs:
```diff
@@ -0,0 +1,76 @@
+        let client = self.client.clone();
+
+        Box::pin(async move {
+            client
+                .publish()
+                .message(entry.message_body)
+                .set_message_group_id(entry.message_group_id)
+                .set_message_deduplication_id(entry.message_deduplication_id)
+                .topic_arn(entry.topic_arn)
```
besides the renames this is one of the biggest changes

neuronull commented:
Hey @wochinge , thanks for proposing this new component. 

Before you go too much further I just wanted to call out we have a procedure to follow for adding new components: https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md#new-sources-sinks-and-transforms

, and in general it starts with a discussion, ideally a GH issue for us to essentially qualify the new component to decide if we'd want to accept it or not.

That said we can still do all of that with the PR state of the changes.

neuronull commented:
> Created an issue @neuronull 👍🏻 Sorry for going the wrong way around - I thought that could maybe make the decision easier.

Hey no worries- it happens a lot. We mostly want to steer towards the q&a first just to avoid folks writing code that might not be utilized.

Thanks for opening that issue and filling out some details there. We'll follow up soon with any thoughts/questions.

neuronull commented on src/sinks/aws_sns/config.rs:
```diff
@@ -0,0 +1,184 @@
+    #[configurable(metadata(
+        docs::examples = "arn:aws:sns:us-east-2:123456789012:MyTopic"
+    ))]
+    pub topic_arn: String,
```
This is the only difference between this sink's config and the SQS config.

We can extract the shared fields into a shared config struct, and use the serde `flatten` to collapse all the shared fields into this one (same on the SQS sink).

neuronull reviewed (changes requested):
Almost all of this code is re-usable with the SQS sink.

We can do something similar to `aws_kinesis_*` sinks (but shouldn't be as complex as that), where there is a common directory that contains the shared code, and then sub directories that contain the integration specific code.

The main (seemingly only) difference is the SNS operates on the topic, and the SQS operates on the queue. These are both strings, so the usage of them should be easy to converge on. And then there is a different Client used from the aws sdk.

Unfortunately it looks like the aws sdk does not seem to utilize a common Trait between the clients, despite them being so similar.
But, we can get around that by defining our own. Basically need a `trait` which has trait functions for any of the function calls to the aws sdk *Client. Then we can use Generics for example in the Service, instead of the `client.publish().<...>` directly in the `call` function, it would look more like

```Rust
    pub(super) struct SnsService<C> {
        pub client: C,
        // ...
    }
     
    impl<C> Service<SendMessageEntry> for SnsService<C>
    where
        C: CommonClientTrait {

    fn call(&mut self, entry: SendMessageEntry) -> Self::Future {
     // ...
     self.client.send(entry).map_ok(|_| SendMessageResponse {
         // ...
 ```

And the SQS and SNS both have an implementation of the trait that handles the differences between them.

Doing that will allow us to eliminate all of the duplicate code that is shared between the components and will be easier to maintain.

I'll hold off on further feedback until then. LMK if my suggestion needs further clarification.

wochinge commented:
Hey @neuronull, thanks for the feedback and the helpful pointers. I have never coded in Rust before but most of the stuff is familiar. Will give it a try in the next days. Another, FYI: I'll be on vacation until September from next Monday onwards. Might be the case that my colleague @ArzelaAscoIi takes over then :-)

wochinge commented:
@neuronull One first question: How do you run the integration tests? It first failed to start the localstack setup and now it insists on running amqp tests 😁 

I used `make test-integration-aws` but also variations like `make test-integration SCOPE="sinks::aws-sqs" ENVIRONMENT=true`

neuronull commented:
Running our integration tests locally can be a bit touchy sometimes. We're loosely aware of this.

Some pointers that could help for future reference: you can utilize the underlying `vdev` commands to have more control over the testing.

When in doubt and you're seeing odd behavior, try `cargo vdev int stop aws` then `cargo vdev int start aws` followed by `cargo vdev int test aws`. The --verbose flag can help with debugging errors. The `test` subcommand is supposed to be a catch all entry point (exercising the same functionality as doing `start` then `test` , but in practice, some of the container dependencies aren't as... robustly defined lets say, as they should be.

wochinge commented on Cargo.toml:
```diff
@@ -771,13 +771,13 @@ amqp-integration-tests = ["sources-amqp", "sinks-amqp"]
-  "aws-kinesis-firehose-integration-tests",
-  "aws-kinesis-streams-integration-tests",
-  "aws-s3-integration-tests",
+#  "aws-cloudwatch-logs-integration-tests",
```
I'll revert these later - this is just to speed up testing

wochinge commented on src/sinks/aws_s_s/sqs/client.rs:
```diff
@@ -0,0 +1,41 @@
+        &self,
+        entry: SendMessageEntry,
+        byte_size: usize,
+    ) -> Result<SendMessageResponse, SdkError<SendMessageError>> {
```
@neuronull Do you have any idea how to abstract the error from this trait? I tried with different variations and I either run into some PhantomData problem (unused generic params) or some variation of generic associated types.  Both are somewhat challenging for someone doing Rust since 6 hours 🙈

neuronull replied:
Is this still relevant? It looks like you overcame this in https://github.com/vectordotdev/vector/pull/18141/commits/891e119e235370b396207629702b0ca1c572eaa5

wochinge replied:
Yep, but if you know any prettier or better suggestion I'd be happy to hear them

wochinge commented:
@neuronull 
* Started with a cleaner implementation. So far focused upon SQS - once the refactor is done adding SNS should be a breeze. I started only with SQS to avoid fighting too many things at once. 
* Happy to already get some initial feedback by you. 
* Last thing missing before we can do SNS is moving out the specific error type which troubles me due to [these issues](https://github.com/vectordotdev/vector/pull/18141#discussion_r1289378391)

wochinge commented on src/sinks/aws_s_s/service.rs:
```diff
@@ -0,0 +1,67 @@
+use super::{client::Client, request_builder::SendMessageEntry};
+
+#[derive(Clone)]
+pub(crate) struct SqsService<C>
```
I of course still need some renaming to do :-)

wochinge commented:
@neuronull Can you maybe have a first look? Did the abstractions as you said, implemented SNS and added a meaningful integration test. Got a bit messy with the Generics though 🙈

neuronull commented:
Side note- if you could avoid force-pushing going forward that would be immensely appreciated 🙏  hard to review incremental changes with force push.

neuronull commented on src/sinks/aws_s_s/sns/config.rs:
```diff
@@ -0,0 +1,114 @@
+
+        let publisher = SnsMessagePublisher::new(client.clone(), self.topic_arn.clone());
+
+        let healthcheck = self.clone().healthcheck(client.clone()).boxed();
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏I know this was from the previous code but I think it's unecessary for the healthcheck to be a part of the struct and require cloning like this. IMO its much cleaner to extract the healthcheck function and pass in the needed info which in this case is just the topic_arn and the client. The Clickhouse sink has an example of this

```Rust
let healthcheck = Box::pin(healthcheck(client.clone(), self.topic_arn.clone());
```

wochinge replied:
done

neuronull commented on src/sinks/aws_s_s/config.rs:
```diff
@@ -102,68 +87,14 @@ pub enum Encoding {
-        &self.acknowledgements
-    }
+#[derive(Clone, Debug)]
+pub struct ConfigWithIds {
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏I think the implementation could be simplified a bit and render this struct not necessary.

This seems to exist only to provide a means for the request builder to call the message_group_id / message_deduplication_id helper functions.
I would suggest instead of having these be helper functions operating on a config struct, extract them as stand alone helper functions that just take the respective fields as input, and the fifo param can be a bool parameter as well.
Those can then be called in the specific `config.rs` files, in the SinkConfig implementations. That `build()` implementation is also where I would recommend instantiating the `RequestBuilder` , which can be passed into the `SSSink::new()` , along with the request (instead of passing the config which is just used to get those).

wochinge replied:
done

neuronull reviewed (commented):
Coming along nicely!

I noticed just a couple of things on the first pass of the refactor. Appreciate the effort put into this. Yeah those generics can get cumbersome 😓

wochinge commented:
Will do 👍🏻 I started the deduplicated version from scratch that's why I needed a force push

neuronull commented on src/sinks/aws_s_s/config.rs:
```diff
@@ -88,13 +88,14 @@ pub enum Encoding {
-pub struct ConfigWithIds {
-    pub base_config: BaseSSSinkConfig,
-    pub fifo: bool,
+pub struct MessageIdConfig {
+    message_group_id: Option<String>,
+    message_deduplication_id: Option<String>,
+    fifo: bool,
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** Ah I should have made my suggestion a bit more clear, I apologize. What I had in mind was removing the need to have a struct like this altogether. So the impl functions of this struct turn into just standalone helper functions.

Something like this:

```Rust

// aws_s_s/config.rs

pub(super) fn message_group_id(message_group_id: Option<String>, fifo: bool)-> crate::Result<Option<Template>> { ... }

pub(super) fn message_deduplication_id(mesage_deduplication_id: Option<String>) -> crate::Result<Option<Template>> { ... }

// aws_s_s/request_builder.rs

impl SSRequestBuilder {
    pub fn new( encoding_config: EncodingConfig, message_group_id: Option<Template>, message_deduplication_id: Option<Template>) -> crate::Result<Self> { ... }

// sns/config.rs

    async fn build(
        &self,
        cx: SinkContext,
    ) -> crate::Result<(crate::sinks::VectorSink, crate::sinks::Healthcheck)> {

        // ...
        
        let request_builder = SSRequestBuilder::new(
            self.base_config.encoding.clone(),
            message_group_id(
                 self.base_config.message_group_id.clone(),
                 self.topic_arn.ends_with(".fifo")
            ),
             message_deduplication_id(self.base_config.message_deduplication_id.clone()
         )?;
         
        let sink = SSSink::new(
            request_builder,
            self.base_config.request.clone(),
            publisher,
        )?;
```

wochinge replied:
done! Btw, @ArzelaAscoIi will take it from here! :-) Thanks for all of your help @neuronull!

ArzelaAscoIi replied:
done :)

ArzelaAscoIi commented on src/sinks/aws_s_s/config.rs:
```diff
@@ -87,41 +87,24 @@ pub enum Encoding {
 
-#[derive(Clone, Debug)]
-pub struct MessageIdConfig {
+pub fn message_group_id(
```
Thanks! Just added a pub(super) 👍

neuronull commented on src/sinks/aws_s_s/sns/mod.rs:
```diff
@@ -1,7 +1,7 @@
 mod integration_tests;
 
-pub use super::{
+pub(super) use super::{
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏So in this and the other file, there isn't really value in having `pub(super) use super::<...>` , because the visibility of the stuff in super, already has visibility to itself. It's redundant essentially. Moreover I didn't notice this earlier but, this actually shouldn't have been `pub` to begin with. The import of the stuff from `super` (the parent module), we only want to access from within this module.

```suggestion
use super::{
```

ArzelaAscoIi replied:
Totally makes sense. Thanks! I replaced `pub` with `pub(super)` where I think it makes sense.

neuronull commented on src/sinks/aws_s_s/config.rs:
```diff
@@ -87,7 +87,7 @@ pub enum Encoding {
 }
 
-pub fn message_group_id(
+pub(super) fn message_group_id(
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏This is a good start but, there are still a bunch of raw `pub` structs, functions, traits, in the changeset that should be `pub(super)`.

neuronull commented:
I'll give this another look shortly.

neuronull commented on src/sinks/aws_s_s/sns/integration_tests.rs:
```diff
@@ -0,0 +1,197 @@
+#![cfg(all(test, feature = "aws-sns-integration-tests"))]
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏This should more appropriately be in mod.rs where the integration_tests module is loaded. (Same for the other one)

neuronull commented on src/sinks/aws_s_s/sns/mod.rs:
```diff
@@ -0,0 +1,12 @@
+    sink::SSSink,
+};
+
+mod client;
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏no reason not to group this up with the rest.

neuronull commented on src/sinks/aws_s_s/sns/config.rs:
```diff
@@ -0,0 +1,125 @@
+/// Configuration for the `aws_sns` sink.
+#[configurable_component(sink(
+    "aws_sns",
+    "Publish observability events to AWS Simple Queue Service topics."
```
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 warning:** ‏

```suggestion
    "Publish observability events to AWS Simple Notification Service topics."
```

neuronull commented on src/sinks/aws_s_s/sqs/mod.rs:
```diff
@@ -0,0 +1,12 @@
+    sink::SSSink,
+};
+
+mod client;
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏can go up with the other modules

neuronull commented on src/sinks/mod.rs:
```diff
@@ -20,8 +20,8 @@ pub mod aws_cloudwatch_metrics;
 pub mod aws_s3;
-#[cfg(feature = "sinks-aws_sqs")]
-pub mod aws_sqs;
+#[cfg(any(feature = "sinks-aws_sqs", feature = "sinks-sinks-aws_sns",))]
```
```suggestion
#[cfg(any(feature = "sinks-aws_sqs", feature = "sinks-sinks-aws_sns"))]
```

neuronull reviewed (dismissed):
This is coming along nicely. Did another pass, what I found was mostly code hygiene nits.

Other things you'll need to do:

`make check-clippy` needs to not return errors
`make generate-component-docs` will create / update cue files for external user facing docs from the changes here, and you then commit the resulting changed files.

ArzelaAscoIi commented on src/sinks/aws_s_s/mod.rs:
```diff
@@ -0,0 +1,15 @@
+pub mod config;
```
done

ArzelaAscoIi commented on src/sinks/aws_s_s/mod.rs:
```diff
@@ -0,0 +1,15 @@
+#[cfg(feature = "sinks-aws_sns")]
+pub mod sns;
+
+mod client;
```
done

ArzelaAscoIi commented:
hi @neuronull. I applied all the changes requests :) 

I just pushed [this commit](https://github.com/vectordotdev/vector/pull/18141/commits/5308c0e5fd75c7000783427a550e87c87bb94e94) to make `make check-clippy` happy. Otherwise I get this error:
<img width="592" alt="image" src="https://github.com/vectordotdev/vector/assets/37148029/1ef80681-33aa-4dc7-bfb7-e867fe578031">

Since `SSSink:new` is used in the sns/sqs sinks (same for the other functions/constructors) these warnings seem to be misleading. However, disabling them seems like I forgot to (maybe) link the parent method somewhere. 


Running `make generate-component-docs` did not create new files locally. Maybe this is connected to the `dead_code` warning 🤔 ?

neuronull commented on src/sinks/aws_s_s/sink.rs:
```diff
@@ -39,6 +39,7 @@ where
     C: Client<E> + Clone + Send + Sync + 'static,
     E: std::fmt::Debug + std::fmt::Display + std::error::Error + Sync + Send + 'static,
 {
+    #[allow(dead_code)]
```
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 warning:** ‏Was this added by mistake? It seems like it shouldn't be necessary since we're using the `new()`function.

Side note- this looks like another `pub` that could be turned into a `pub(super)`

neuronull commented on src/sinks/aws_s_s/sns/mod.rs:
```diff
@@ -1,3 +1,6 @@
+#![cfg(all(test, feature = "aws-sns-integration-tests"))]
```
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 warning:** ‏This should actually go right above the line `mod integration_tests;`. The way the macro works is it gates the scope of the module it is declared over.

neuronull commented on src/sinks/aws_s_s/sqs/mod.rs:
```diff
@@ -1,3 +1,6 @@
+#![cfg(all(test, feature = "aws-sqs-integration-tests"))]
```
(same)

neuronull reviewed (commented):
Just a couple follow-ups

neuronull commented:
> 
> Since `SSSink:new` is used in the sns/sqs sinks (same for the other functions/constructors) these warnings seem to be misleading. However, disabling them seems like I forgot to (maybe) link the parent method somewhere.
> 
> Running `make generate-component-docs` did not create new files locally. Maybe this is connected to the `dead_code` warning 🤔 ?

It's quite likely related to the comment I raised about the macro to gate integration tests, I'd try moving those and then checking again.

ArzelaAscoIi commented:
You were totally right. It was caused by the macro placed on top 🤦 Thanks! Currently running `make generate-component-docs` again. Will push the docs afterwards :)

neuronull reviewed (approved):
Awesome, triggering CI

ArzelaAscoIi commented on src/sinks/aws_s_s/config.rs:
```diff
@@ -0,0 +1,102 @@
+#[serde(deny_unknown_fields)]
+pub(super) struct BaseSSSinkConfig {
+    #[serde(flatten)]
+    pub(super) region: RegionOrEndpoint,
```
I moved the region down to the sqs/sns and ran the commands you suggested, but added the `cargo run --bin` prefix to use current branch. 

```
cargo run --bin vector generate stdin//aws_sns > config.toml
cargo run --bin  vector validate config.toml 
```
It seems like the path is connected to my setup, but generating seems to work. This was the output:
```sh
❯ cargo run --bin  vector validate config.toml                

    Finished dev [unoptimized + debuginfo] target(s) in 1.17s
     Running `target/debug/vector validate config.toml`
√ Loaded ["config.toml"]
x data_dir "/var/lib/vector/" does not exist
```

neuronull replied:
ah, yeah that would happen if that dir doesnt exist on the filesystem or doesnt have permissions. can get around that by editing the data_dir field in the config to something like `/tmp/vector/` . also probably have to disable the healthcheck. anyway, the change looks good so I'll re-run CI.

neuronull commented:
Thanks for your contribution @wochinge and @ArzelaAscoIi  !

ArzelaAscoIi commented:
Hi @neuronull. Do you know roughly how long it would take to release the SNS sink ?

jszwedko commented:
> Hi @neuronull. Do you know roughly how long it would take to release the SNS sink ?

The next release is targeted for the 26th. You can run one of the nightly builds if you want to try it out earlier though.
